### ISSUE TITLE
Adjust XrdCl operation timeouts based on forward progress

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#2333]
CMS observes that vector read timeouts are one of the most common failure modes.  Looking at the frequency of success on retries, it appears that the vector read was progressing fine -- however, the timeout (which is per-operation) is simply too small for a large, complex vector read.

This results in a conundrum: set the timeout large enough for the vector read, and it's way too large for a single-read (you stall the process a long time for really broken servers).  Set it small enough for a single-read and the vector read errors out frequently. 
 @smithdh noted that timeouts could be adjusted by CMS per-operation; that's certainly true but then no one else would benefit.

Trying to resize the timeout based on request size likely isn't the best approach: if your heuristic determines a very large request should have a 5 minute timeout, there may be zero forward progress for the entire 5 minutes.  Heuristics are also hard for something like a large dirlist request: we don't know the size of the request until the request has finished.

So, here's a simple proposal:

Add a "progress timeout" alongside the "operation timeout" (though, for simplicity, we might consider reinterpreting the operation timeout as a progress timeout).  The progress timeout counter should be reset each time partial results are received from the server.

That is, if the progress timeout is 10s, then each time a chunk of a directory listing (or a read request or a vector read request) is received, reset the timeout to 10s into the future.  With this proposed setup, a 10s progress timeout only triggers when the server has provided no additional data in the last 10s.

**Comment by abh3** [comment]
@bbockelm so I had a long discussion about this with @amadio. First, as you pointed out, this issue affects everyt operation not just readv. So, the solution needs to be global. Second, we do have a progress timeout already. It is called XRD_STREAMTIMEOUT (-DIStreamTimeout) which defaults to 60 seconds. Is a socket sees no activity in 60 seconds after a request has been sent, it is considered an error. When there is activity, the timeout is reset. So, this is exactly a progress timeout. The other timeout is XRD_REQUESTTIMEOUT (-DIRequestTimeout) which defaults to 1800 seconds (30 mnutes).

So, based on the description of the problem, I think the real issue is that a stream timeout of 60 seconds is too small (it's pretty rare to hit the request timeout). I seriously doubt anyone wants a request to continue for 30 minutes and this need a request timeout but some find that 60 seconds is too small (e.g. CTA) which may be the problem here. However, 60 seconds is a pretty long time for a server not to make forward progress. So, perhaps something else is going on and we really should be looking elsewhere. For instance, it could be that someone adjusted the values (likely via envar) to a very small number of seconds and that is the root cause of the problem you are seeing.

In any case, given that we already have a progress timeout along with a very generous request timeout, I don't think we need to make any changes, do you at this point?

**Comment by adriansev** [comment]
Hi! As i'm interested in the subject i will chime in: i think that there is a distinction between "Operation X was not completed in given Y time" and "Bytes flowed in the last X seconds so lets not abandon this" (or "No bytes in the last X seconds, let's abandon this operation")
In principle there is XRateThreshold that already do this but is doing it's computation at the end of chunk so with default 8MiB (and 4 parallel chunks) and if there is a problem (you have either 0 bytes/s or 100 bytes/s) you will have to wait for first chunk to get a bytes/s value that then will take action if you specified your rate threshold .. 
It would be very interesting a benchmark with sizes and number of chunks against an unloaded server and against a loaded one (for both ssd and hdd) as maybe smaller chunks (but more of them) would help with faster decision if the connection should be abandoned or not. 
I think that what Brian expressed is something _during_ the chunk transfer (so the chunk did not finished) to be able to do something like "ok, the timeout is not reached but in the last X seconds the amount Y of bytes flown is to small, so let's bail out" and for the positive case "ok, it is taking a lot of time, but at least enough Y bytes passed in the last X seconds, so let's continue" (i hope that i understood correctly Brian's message)

**Comment by bbockelm** [comment]
@abh3 - you write:

> If a socket sees no activity in 60 seconds after a request has been sent, it is considered an error

Are you sure that's correct?  I was reading through the code and it seems that the stream timeout fires 60 seconds after a SID has been allocated; I didn't see any indication the SID is re-allocated in the client after receiving a partial message.  (But this part of the code is hard to trace -- would appreciate knowing if there's any test coverage that it works like we think it does)

Using the stream timeout also switches the error condition from `errOperationExpired` to `errSocketTimeout`.  I'd have to think about what that implies for things like retries and client behaviors.  Any insights?

**Comment by amadio** [comment]
> would appreciate knowing if there's any test coverage that it works like we think it does

I don't think we have, we will add a test for this, it shouldn't be too hard to modify one of the current tests to check this.

**Comment by amadio** [comment]
Related issue: #1259
